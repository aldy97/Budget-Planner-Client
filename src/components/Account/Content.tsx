import React, { useState } from "react";
import { Layout, Input, Space, Button } from "antd";
import { message } from "antd";
import axios from "axios";
import Progress from "./Progress";
import { UPDATE_USER_INFO, UpdateUserInfo } from "../../actions/HomeAction";
import { User } from "../../reducers/HomeReducer";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import { RootState } from "../../reducers/index";

interface ContenProps {
  user: User;
  updateUserInfo: (user: User) => void;
}

function Content({ user, updateUserInfo }: ContenProps) {
  const { Content } = Layout;
  const [currBudget, setCurrBudget] = useState<number>(user.budget);
  const [currThreshold, setCurrThrehold] = useState<number>(user.threshold);

  const handleBudgetChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setCurrBudget(parseInt(e.target.value));
  };

  const handleConfirmBtnClick = async () => {
    if (currBudget < 0) {
      message.error("Budget must be greater than 0");
      return;
    }
    const request = {
      _id: user._id,
      updatedFields: { budget: currBudget, threshold: currThreshold },
    };
    const response = await axios.put("/api/updateUserInfo", request);
    if (response.data.status) {
      message.success(response.data.message);
    } else {
      message.error(response.data.message);
    }
  };

  return (
    <Content style={{ margin: "4px 16px" }}>
      <div
        className="site-layout-background"
        style={{ padding: 30, minHeight: 360, marginTop: 30 }}
      >
        <Space direction="vertical" size="large">
          <div>
            <div>Name:</div>
            <Input defaultValue={user.name} disabled data-test="name"></Input>
          </div>
          <div>
            <div>Email:</div>
            <Input defaultValue={user.email} style={{ width: 224 }} disabled></Input>
          </div>
          <div>
            <div>Set monthly budget:</div>
            <Input
              defaultValue={currBudget}
              prefix="$"
              suffix="CAD"
              onChange={handleBudgetChange}
            />
          </div>
          {/* <div>
            <div>New Password</div>
            <Input.Password
              iconRender={visible =>
                visible ? <EyeTwoTone /> : <EyeInvisibleOutlined />
              }
            ></Input.Password>
          </div>
          <div>
            <div>Confirm Password</div>
            <Input.Password
              iconRender={visible =>
                visible ? <EyeTwoTone /> : <EyeInvisibleOutlined />
              }
            ></Input.Password>
          </div> */}
          <div>
            <div>Set reminder for my budget:</div>
            <Progress
              threshold={currThreshold}
              changeThreshold={setCurrThrehold}
            ></Progress>
          </div>
          <Button type="primary" onClick={handleConfirmBtnClick}>
            Confirm
          </Button>
        </Space>
      </div>
    </Content>
  );
}

const mapState = (state: RootState) => {
  return {
    user: state.HomeReducer.user,
  };
};

const mapDispatch = (dispatch: Dispatch) => {
  return {
    updateUserInfo(user: User) {
      const action: UpdateUserInfo = {
        type: UPDATE_USER_INFO,
        user,
      };
      dispatch(action);
    },
  };
};

export default connect(mapState, mapDispatch)(Content);
