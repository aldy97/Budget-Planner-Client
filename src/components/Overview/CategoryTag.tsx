import React from "react";
import { COLORS } from "../../utils/constants";
import { Tag } from "antd";

interface TagProps {
  category: string;
  type: string;
}

function CategoryTag({ category, type }: TagProps) {
  return (
    <Tag color={type === "expense" ? COLORS.ORANGE : COLORS.THEMEBLUE}>
      {category}
    </Tag>
  );
}

export default CategoryTag;
