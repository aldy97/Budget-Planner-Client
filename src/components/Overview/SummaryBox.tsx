import React from "react";
import styled from "styled-components";

const StyledBox = styled.div<Box>`
  padding: 80px;
  border-radius: 10px;
  margin: 32px;
  text-align: center;
  flex: 1;
  font-size: 30px;
  background: ${props =>
    props.type === "income"
      ? "linear-gradient(rgb(66, 161, 236), rgb(0, 112, 201))"
      : "linear-gradient(#ffa940,#d46b08)"};
  color: #fff;
`;

interface Box {
  type?: "expense" | "income";
  amount?: string;
}

function SummaryBox({ type, amount }: Box): JSX.Element {
  return (
    <StyledBox type={type}>
      <div data-test="type">
        {type === "expense" ? "Monthly Expense" : "Monthly Income"}:
      </div>
      <div data-test="amount">${amount}</div>
    </StyledBox>
  );
}

export default SummaryBox;
