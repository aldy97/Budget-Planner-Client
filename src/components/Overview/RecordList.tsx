import React from "react";
import Tag from "./CategoryTag";
import { Record } from "./Content";
import { List } from "antd";
import moment from "moment";

interface ListProps {
  type: "expense" | "income";
  records: Record[];
  maxLength: number;
}

function RecordList({ type, records, maxLength }: ListProps): JSX.Element {
  const currMonth = moment().month() + 1;
  const currYear = moment().year();

  const matchYearAndMonth = (date: string): boolean => {
    const [year, month] = date.split("-");
    return parseInt(year) === currYear && parseInt(month) === currMonth;
  };

  // 根据记录类型，展示当年当月指定数量的记录
  const data = records
    .filter(record => record.type === type)
    .filter(record => matchYearAndMonth(record.recordDate))
    .reverse()
    .slice(0, maxLength);

  return (
    <>
      <List
        style={{ flex: 1, marginLeft: 30, marginRight: 30 }}
        itemLayout="horizontal"
        dataSource={data}
        renderItem={item => (
          <List.Item data-test="list-item">
            <List.Item.Meta
              title={item.title}
              description={
                <div>
                  <Tag type={item.type} category={item.category}></Tag>
                  <div>{item.description}</div>
                </div>
              }
            />
            <div style={{ marginRight: 30, color: "#8c8c8c" }}>${item.amount}</div>
          </List.Item>
        )}
      />
    </>
  );
}

export default RecordList;
