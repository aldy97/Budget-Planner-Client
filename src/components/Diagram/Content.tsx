import React, { useState, useEffect } from "react";
import styled from "styled-components";
import { Layout, Dropdown, Button, Menu, Space } from "antd";
import LineChart from "./LineChart";
import PieChart from "./PieChart";
import { Record } from "../Overview/Content";
import { connect } from "react-redux";
import { RootState } from "../../reducers/index";
import moment from "moment";

const DropdownWrapper = styled.div`
  text-align: center;
`;

const StyledChartsWrapper = styled.div`
  display: flex;
`;

const LeftSection = styled.div`
  flex: 1;
`;

const RightSection = styled.div`
  flex: 1;
`;

const Title = styled.div`
  text-align: center;
  color: #595959;
  font-size: 20px;
  font-weight: bold;
`;

export enum Options {
  LAST_WEEK = "In the past week",
  LAST_MONTH = "In the past month",
}

interface ContentProps {
  records?: Record[];
}

function Content({ records }: ContentProps): JSX.Element {
  const { Content } = Layout;
  const [period, setPeriod] = useState<string>(Options.LAST_WEEK);
  const [options, setOptions] = useState<string[]>([]);

  const initOptions = (): void => {
    let tempOptions: string[] = [];
    tempOptions = Object.values(Options);

    const dates = records?.map(record =>
      moment(record.recordDate).format("YYYY-MM")
    );

    for (const date of dates as string[]) {
      if (!tempOptions.includes(date)) {
        tempOptions.push(date);
      }
    }

    setOptions(tempOptions);
  };

  // when records change, options will also change accordingly
  useEffect(() => {
    initOptions();
  }, [records]);

  const menu = (
    <Menu>
      {options.map(option => {
        return (
          <Menu.Item
            key={option}
            onClick={() => {
              setPeriod(option);
            }}
          >
            <a>{option}</a>
          </Menu.Item>
        );
      })}
    </Menu>
  );

  return (
    <Content style={{ margin: "4px 16px" }}>
      <div
        className="site-layout-background"
        style={{ padding: 30, minHeight: 360, marginTop: 30 }}
      >
        <DropdownWrapper>
          <Space direction="horizontal">
            <Dropdown overlay={menu} placement="bottomCenter">
              <Button>{period}</Button>
            </Dropdown>
          </Space>
        </DropdownWrapper>
        <StyledChartsWrapper>
          <LeftSection>
            <Title>Expense Distribution</Title>
            <PieChart
              records={records as Record[]}
              type="expense"
              period={period}
            ></PieChart>
          </LeftSection>
          <RightSection>
            <Title>Income Distribution</Title>
            <PieChart
              records={records as Record[]}
              type="income"
              period={period}
            ></PieChart>
          </RightSection>
        </StyledChartsWrapper>
        <div style={{ marginLeft: 120, marginRight: 120 }}>
          <Title>Expense Trend</Title>
          <LineChart
            records={records as Record[]}
            type="expense"
            period={period}
          ></LineChart>
          <Title>Income Trend</Title>
          <LineChart
            records={records as Record[]}
            type="income"
            period={period}
          ></LineChart>
        </div>
      </div>
    </Content>
  );
}

const mapState = (state: RootState) => {
  return {
    records: state.HomeReducer.records,
  };
};

export default connect(mapState, null)(Content);
