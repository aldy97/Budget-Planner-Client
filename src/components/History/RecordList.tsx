import React, { useState, useEffect } from "react";
import { User } from "../../reducers/HomeReducer";
import ListItemMeta from "./ListItemMeta";
import { COLORS } from "../../utils/constants";
import { EditOutlined, DeleteOutlined } from "@ant-design/icons";
import { Record } from "../Overview/Content";
import axios from "axios";
import { List, message, Popconfirm, Button } from "antd";
import { connect } from "react-redux";
import { RootState } from "../../reducers/index";
import { Dispatch } from "redux";
import { UpdateRecords, UPDATE_RECORDS } from "../../actions/HomeAction";
import { Filter } from "../../reducers/FilterReducer";

interface List {
  user?: User;
  records?: Record[];
  filter?: Filter;
  updateRecordsToRedux?: (records: Record[]) => void;
}

interface UpdateRequest {
  _id: string;
  updatedFields: {
    [key: string]: string | number;
  };
}

const dummySelected: Record = {
  _id: "string",
  amount: 0,
  description: "",
  title: "",
  type: "",
  category: "",
  recordDate: "",
  createdOn: "",
  updatedOn: "",
};

function RecordList({ user, records, filter, updateRecordsToRedux }: List) {
  const recordList = records || [];
  const currUser = user ? user : null;
  const [data, setData] = useState<Record[]>(recordList);
  const [selected, setSelected] = useState<Record>(dummySelected);

  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [amount, setAmount] = useState(0);
  const [recordDate, setRecordDate] = useState("");

  const [delID, setDelID] = useState("");

  useEffect(() => {
    setData(recordList);
  }, [records]);

  useEffect(() => {
    setTitle(selected.title);
    setDescription(selected.description);
    setAmount(selected.amount);
    setRecordDate(selected.recordDate);
  }, [selected]);

  const getSortData = (records: Record[]): Record[] => {
    const sortedData = records.sort((a, b) => {
      if (a.recordDate < b.recordDate) {
        return 1;
      }
      return a.amount >= b.amount ? 1 : -1;
    });
    return sortedData;
  };

  const generateRecords = (): void => {
    let modifiedRecord: Record[] = [];
    const enabled = filter?.enabled;
    const month = filter?.month;
    const category = filter?.category;
    if (!enabled || (month === "" && category === "")) {
      modifiedRecord = recordList;
    } else {
      if (month && category && month !== "" && category !== "") {
        modifiedRecord = recordList.filter(
          record =>
            record.category === category && record.recordDate.slice(0, 7) === month
        );
      } else if (!month && !category) {
        modifiedRecord = recordList;
      } else if (month && month !== "") {
        modifiedRecord = recordList.filter(
          record => record.recordDate.slice(0, 7) === month
        );
      } else {
        modifiedRecord = recordList.filter(record => record.category === category);
      }
    }
    // caution: update state only at the top level
    setData(getSortData(modifiedRecord));
  };

  useEffect(() => {
    generateRecords();
  }, [filter]);

  const updateAllRecordsToRedux = async (): Promise<void> => {
    const response = await axios.get(`/api/getRecords/${currUser?._id}`);
    const records: Record[] = response.data.records;
    if (updateRecordsToRedux) {
      updateRecordsToRedux(records);
      console.log("updated");
    }
  };

  const onDeleteConfirm = async (recordID: string): Promise<void> => {
    const request = { data: { recordID } };
    const response = await axios.delete("/api/deleteRecord", request);
    console.log(response);
    if (response.status === 202) {
      updateAllRecordsToRedux();
      message.success(response.data.message);
    } else {
      message.error(response.data.message);
    }
  };

  const handleEditBtnClick = (record: Record): void => {
    if (record._id !== selected._id) {
      setSelected(record);
    } else {
      setSelected(dummySelected);
    }
  };

  const handleConfirmEditBtnClick = async (): Promise<void> => {
    if (recordDate === "") {
      message.warn("Please select a record date");
      return;
    }
    if (amount === 0) {
      message.warn("Amount must be greater than 0");
      return;
    }

    const request: UpdateRequest = {
      _id: selected._id,
      updatedFields: {
        title: title || "No title",
        amount,
        recordDate,
        description: description || "No description",
      },
    };
    const response = await axios.put("/api/updateRecord", request);

    if (response.status === 200) {
      message.success(response.data.message);
      updateAllRecordsToRedux();
      setSelected(dummySelected);
    } else {
      message.error(response.data.message);
    }
  };

  return (
    <>
      <List
        style={{ flex: 1 }}
        itemLayout="horizontal"
        dataSource={getSortData(data)}
        renderItem={item => (
          <List.Item
            key={item._id}
            actions={[
              <EditOutlined
                onClick={() => {
                  handleEditBtnClick(item);
                }}
                style={{ color: COLORS.THEMEBLUE, cursor: "pointer" }}
                key="edit-item"
              ></EditOutlined>,
              <Button
                disabled={selected._id !== item._id}
                type="primary"
                key="confirm-edit"
                size="small"
                onClick={handleConfirmEditBtnClick}
              >
                Confirm Edit
              </Button>,
              <Popconfirm
                key="delete-item"
                placement="topLeft"
                title="Are you sure you want to delete this record?"
                onConfirm={() => {
                  onDeleteConfirm(item._id);
                }}
                okText="Yes"
                cancelText="No"
              >
                <DeleteOutlined
                  onClick={() => {
                    setDelID(item._id);
                  }}
                  style={{ color: "#f5222d", cursor: "pointer" }}
                ></DeleteOutlined>
              </Popconfirm>,
            ]}
          >
            <ListItemMeta
              selected={selected}
              item={item}
              setTitle={setTitle}
              setDescription={setDescription}
              setAmount={setAmount}
              setRecordDate={setRecordDate}
              data-test="list-item"
            ></ListItemMeta>
          </List.Item>
        )}
      />
    </>
  );
}

const mapState = (state: RootState) => {
  return {
    user: state.HomeReducer.user,
    records: state.HomeReducer.records,
    filter: state.FilterReducer.filter,
  };
};

const mapDispatch = (dispatch: Dispatch) => {
  return {
    updateRecordsToRedux(records: Record[]) {
      const action: UpdateRecords = {
        type: UPDATE_RECORDS,
        records,
      };
      dispatch(action);
    },
  };
};

export default connect(mapState, mapDispatch)(RecordList);
