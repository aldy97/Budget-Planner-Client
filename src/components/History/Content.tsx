import React from "react";
import { Layout } from "antd";
import Filter from "./Filter";
import RecordList from "./RecordList";

function Content() {
  const { Content } = Layout;
  return (
    <Content style={{ margin: "4px 16px" }}>
      <div
        className="site-layout-background"
        style={{
          paddingLeft: 152,
          paddingRight: 152,
          paddingTop: 24,
          paddingBottom: 24,
          minHeight: 360,
          marginTop: 30,
        }}
      >
        <div style={{ textAlign: "center" }}>
          <Filter></Filter>
        </div>
        <RecordList></RecordList>
      </div>
    </Content>
  );
}

export default Content;
