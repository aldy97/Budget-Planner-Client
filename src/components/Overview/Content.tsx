import React, { useEffect, useState } from "react";
import styled from "styled-components";
import SummaryBox from "./SummaryBox";
import RecordList from "./RecordList";
import { Layout } from "antd";
import { connect } from "react-redux";
import { RootState } from "../../reducers/index";

const SummaryWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

const ListsWrapper = styled.div`
  display: flex;
`;

const StyledHint = styled.div`
  margin-top: 10px;
  text-align: center;
  color: #8c8c8c;
`;

export interface Record {
  _id: string;
  amount: number;
  description: string;
  title: string;
  type: string;
  category: string;
  recordDate: string;
  createdOn: string;
  updatedOn: string;
}

interface ContentProps {
  records?: Record[];
}

function Content({ records }: ContentProps) {
  const { Content } = Layout;

  const [expenseMonthly, setExpenseMonthly] = useState(0);
  const [incomeMonthly, setIncomeMonthly] = useState(0);

  const getExpenseAndIncome = () => {
    let expense = 0;
    let income = 0;
    const expenseRecords: Record[] = records
      ? records.filter(record => record.type === "expense")
      : [];
    const incomeRecords: Record[] = records
      ? records.filter(record => record.type === "income")
      : [];
    expenseRecords.map(records => (expense += records.amount));
    incomeRecords.map(records => (income += records.amount));
    setExpenseMonthly(expense);
    setIncomeMonthly(income);
  };

  useEffect(() => {
    getExpenseAndIncome();
  }, [records]);

  return (
    <Content style={{ margin: "4px 16px" }}>
      <div
        className="site-layout-background"
        style={{ padding: 24, paddingTop: 5, minHeight: 360, marginTop: 30 }}
      >
        <SummaryWrapper>
          <SummaryBox type="expense" amount={expenseMonthly}></SummaryBox>
          <SummaryBox type="income" amount={incomeMonthly}></SummaryBox>
        </SummaryWrapper>
        <ListsWrapper>
          <RecordList
            records={records as Record[]}
            maxLength={5}
            type="expense"
          ></RecordList>
          <RecordList
            records={records as Record[]}
            maxLength={5}
            type="income"
          ></RecordList>
        </ListsWrapper>
        <StyledHint>Only 5 most recent records are shown</StyledHint>
      </div>
    </Content>
  );
}

const mapStateToProps = (state: RootState) => {
  return {
    records: state.HomeReducer.records,
  };
};

export default connect(mapStateToProps, null)(Content);
