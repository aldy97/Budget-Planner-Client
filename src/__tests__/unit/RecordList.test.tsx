import React from "react";
import { mount } from "enzyme";
import { Record } from "../../components/Overview/Content";
import RecordList from "../../components/Overview/RecordList";

const mockRecords: Record[] = [
  {
    title: "test1",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2021-01-01",
  },
  {
    title: "test2",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2021-01-01",
  },
  {
    title: "test3",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2021-01-01",
  },
  {
    title: "test4",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2021-01-01",
  },
  {
    title: "test5",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2021-01-01",
  },
  {
    title: "test6",
    description: "No description",
    createdOn: "2020-01-01T00:00:00.000Z",
    updatedOn: "2020-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "expense",
    category: "Meals",
    amount: 20,
    recordDate: "2019-01-01",
  },
  {
    title: "test7",
    description: "No description",
    createdOn: "2021-01-01T00:00:00.000Z",
    updatedOn: "2021-01-01T00:00:00.000Z",
    _id: "5fef5ab6f90ada2dfaa37482",
    type: "income",
    category: "investment",
    amount: 20,
    recordDate: "2019-01-01",
  },
];

describe("RecordList", () => {
  const wrapper = mount(
    <RecordList records={mockRecords} maxLength={5} type="expense"></RecordList>
  );

  it("renders porperly", () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("receives correct number of records", () => {
    expect(wrapper.prop("records").length).toBe(7);
  });

  it("only renders 5 records in current year and month under given category", () => {
    const listItems = wrapper.find("li");
    expect(listItems.length).toBe(5);
    expect(listItems.at(0).find("h4").text()).toEqual("test5");
    expect(listItems.at(1).find("h4").text()).toEqual("test4");
    expect(listItems.at(2).find("h4").text()).toEqual("test3");
    expect(listItems.at(3).find("h4").text()).toEqual("test2");
    expect(listItems.at(4).find("h4").text()).toEqual("test1");
  });
});
