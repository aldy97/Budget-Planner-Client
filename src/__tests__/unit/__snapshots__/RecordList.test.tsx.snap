// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordList renders porperly 1`] = `ReactWrapper {}`;
